pkg/protocol/tipmanager/v1/tipmanager.go

```diff
@@ -0,0 +1,363 @@
+	optMaxLikedInsteadReferences int
+
+	// optMaxLikedInsteadReferencesPerParent contains the maximum number of liked instead references that are allowed
+	// per parent.
+	optMaxLikedInsteadReferencesPerParent int
+
+	// optMaxWeakReferences contains the maximum number of weak references that are allowed.
+	optMaxWeakReferences int
+
+	module.Module
+}
+
+// NewProvider creates a new TipManager provider.
+func NewProvider(opts ...options.Option[TipManager]) module.Provider[*engine.Engine, tipmanager.TipManager] {
+	return module.Provide(func(e *engine.Engine) tipmanager.TipManager {
+		t := NewTipManager(e.Ledger.ConflictDAG(), e.BlockCache.Block, e.EvictionState.LatestRootBlocks, opts...)
+
+		e.Events.Booker.BlockBooked.Hook(t.AddBlock, event.WithWorkerPool(e.Workers.CreatePool("AddTip", 2)))
+		e.BlockCache.Evict.Hook(t.Evict)
+		e.HookStopped(t.Shutdown)
+		e.Events.TipManager.LinkTo(t.events)
+
+		t.TriggerInitialized()
+
+		return t
+	})
+}
+
+// NewTipManager creates a new TipManager.
+func NewTipManager(conflictDAG conflictdag.ConflictDAG[iotago.TransactionID, iotago.OutputID, booker.BlockVotePower], blockRetriever func(blockID iotago.BlockID) (block *blocks.Block, exists bool), rootBlocksRetriever func() iotago.BlockIDs, opts ...options.Option[TipManager]) *TipManager {
+	return options.Apply(&TipManager{
+		retrieveBlock:                blockRetriever,
+		retrieveRootBlocks:           rootBlocksRetriever,
+		conflictDAG:                  conflictDAG,
+		tipMetadataStorage:           shrinkingmap.New[iotago.SlotIndex, *shrinkingmap.ShrinkingMap[iotago.BlockID, *TipMetadata]](),
+		strongTipSet:                 randommap.New[iotago.BlockID, *TipMetadata](),
+		weakTipSet:                   randommap.New[iotago.BlockID, *TipMetadata](),
+		events:                       tipmanager.NewEvents(),
+		optMaxStrongParents:          8,
+		optMaxLikedInsteadReferences: 8,
+		optMaxWeakReferences:         8,
+	}, opts, func(t *TipManager) {
+		t.optMaxLikedInsteadReferencesPerParent = t.optMaxLikedInsteadReferences / 2
+	})
+}
+
+// AddBlock adds a Block to the TipManager.
+func (t *TipManager) AddBlock(block *blocks.Block) {
+	newBlockMetadata := NewBlockMetadata(block)
+
+	if storage := t.metadataStorage(block.ID().Index()); storage != nil && storage.Set(block.ID(), newBlockMetadata) {
+		t.setupBlockMetadata(newBlockMetadata)
+	}
+}
+
+// SelectTips selects the references that should be used for block issuance.
+func (t *TipManager) SelectTips(amount int) (references model.ParentReferences) {
+	references = make(model.ParentReferences)
+
+	seenStrongTips := advancedset.New[iotago.BlockID]()
+	selectStrongTips := func(amount int) (strongTips []*TipMetadata) {
+		if amount > 0 {
+			for _, strongTip := range t.strongTipSet.RandomUniqueEntries(amount + seenStrongTips.Size()) {
+				if seenStrongTips.Add(strongTip.Block().ID()) {
+					if strongTips = append(strongTips, strongTip); len(strongTips) == amount {
```

Comment:
I am not sure I understand what you mean. We use this pattern all over the place:

if myValue := determineValue(); myValue == someExpectedValue {
    // do sth
}

Why is this bad? It just assigns the variable and then checks the value?

You means because its also used as a return value? I am not sure I get the concern but we can put the call a line on top :P

Reply:
I am not sure I understand what you mean. We use this pattern all over the place:

```
if myValue := determineValue(); myValue == someExpectedValue {
    // do sth
}
```

Why is this bad? It just assigns the variable and then checks the value?

You means because its also used as a return value? I am not sure I get the concern but we can put the call a line on top :P